crates/sui-types/src/multisig.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    crypto::{CompressedSignature, SignatureScheme, SuiSignature},
+    serde_to_from_bytes,
+    sui_serde::SuiBitmap,
+};
+pub use enum_dispatch::enum_dispatch;
+use fastcrypto::{
+    ed25519::Ed25519PublicKey,
+    encoding::Base64,
+    error::FastCryptoError,
+    secp256k1::Secp256k1PublicKey,
+    secp256r1::Secp256r1PublicKey,
+    traits::{EncodeDecodeBase64, ToFromBytes},
+    Verifier,
+};
+use once_cell::sync::OnceCell;
+use roaring::RoaringBitmap;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use std::hash::{Hash, Hasher};
+
+use crate::{
+    base_types::SuiAddress,
+    crypto::{PublicKey, Signature},
+    error::SuiError,
+    intent::IntentMessage,
+};
+
+#[cfg(test)]
+#[path = "unit_tests/multisig_tests.rs"]
+mod multisig_tests;
+
+pub type WeightUnit = u8;
+pub type ThresholdUnit = u16;
+pub const MAX_SIGNER_IN_MULTISIG: usize = 10;
+
+/// A lightweight trait that all members of [enum GenericSignature] implement.
+#[enum_dispatch]
+pub trait AuthenticatorTrait {
+    fn verify_secure_generic<T>(
+        &self,
+        value: &IntentMessage<T>,
+        author: SuiAddress,
+    ) -> Result<(), SuiError>
+    where
+        T: Serialize;
+}
+
+/// Due to the incompatibility of [enum Signature] (which dispatches a trait that
+/// assumes signature and pubkey bytes for verification), here we add a wrapper
+/// enum where member can just implement a lightweight [trait AuthenticatorTrait].
+/// This way multisig (and future Authenticators) can implement its own `verify`.
+#[enum_dispatch(AuthenticatorTrait)]
+#[derive(Debug, Clone, PartialEq, Eq, JsonSchema, Hash)]
+pub enum GenericSignature {
+    MultiSignature,
+    Signature,
+}
+
+/// GenericSignature encodes a single signature [enum Signature] as is `flag || signature || pubkey`.
+/// It encodes [struct MultiSignature] as the multisig flag (0x03) concat with the bcs serializedbytes
+/// of [struct MultiSignature] i.e. `flag || bcs_bytes(multisig)`.
+impl ToFromBytes for GenericSignature {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FastCryptoError> {
+        match SignatureScheme::from_flag_byte(
+            bytes.first().ok_or(FastCryptoError::InputTooShort(0))?,
+        ) {
+            Ok(x) => match x {
+                SignatureScheme::Multisig => {
+                    let multisig: MultiSignature =
+                        bcs::from_bytes(bytes.get(1..).ok_or(FastCryptoError::InvalidInput)?)
+                            .map_err(|_| FastCryptoError::InvalidSignature)?;
+                    Ok(GenericSignature::MultiSignature(multisig))
+                }
+                SignatureScheme::Secp256k1
```

Comment:
I'd put those in the correct order (as we have defined them in flags) for readability